File: MasteringKotlin/app/src/main/java/com/devlokos/masteringkotlin/chapters/chapter09/intermediate/IntermediateInheritance.kt
package com.devlokos.masteringkotlin.chapters.chapter09.intermediate

/**
 * Capítulo 09 — Herencia y Polimorfismo.
 * Nivel INTERMEDIO: Clase base Vehicle, subclases Car y Bike; startEngine() sobrescrito.
 */
fun main() {
    runIntermediateInheritanceExamples()
}

/** Clase base para vehículos. */
open class Vehicle(val brand: String) {
    open fun startEngine() {
        println("$brand vehicle is starting...")
    }
}

/** Car sobrescribe startEngine(). */
class Car(brand: String) : Vehicle(brand) {
    override fun startEngine() {
        println("$brand car: turning the key, engine running.")
    }
}

/** Bike sobrescribe startEngine(). */
class Bike(brand: String) : Vehicle(brand) {
    override fun startEngine() {
        println("$brand bike: kick-start, engine running.")
    }
}

fun runIntermediateInheritanceExamples() {
    val car = Car("Toyota")
    val bike = Bike("Harley")
    car.startEngine()
    bike.startEngine()
}

File: MasteringDart/lib/chapters/chapter09/intermediate/intermediate_inheritance.dart
// ignore_for_file: avoid_print
//
//  intermediate_inheritance.dart
//  MasteringDart
//
//  Capítulo 09 — Herencia y polimorfismo.
//  Nivel INTERMEDIO: Vehicle, Car, Bike; startEngine() sobrescrito.
//
//  Ejecutar con: dart run lib/chapters/chapter09/intermediate/intermediate_inheritance.dart

/// Clase base para vehículos.
class Vehicle {
  final String brand;
  Vehicle(this.brand);
  void startEngine() {
    print('$brand vehicle is starting...');
  }
}

/// Car sobrescribe startEngine().
class Car extends Vehicle {
  Car(super.brand);
  @override
  void startEngine() {
    print('$brand car: turning the key, engine running.');
  }
}

/// Bike sobrescribe startEngine().
class Bike extends Vehicle {
  Bike(super.brand);
  @override
  void startEngine() {
    print('$brand bike: kick-start, engine running.');
  }
}

void runIntermediateInheritanceExamples() {
  final car = Car('Toyota');
  final bike = Bike('Harley');
  car.startEngine();
  bike.startEngine();
}

void main() {
  runIntermediateInheritanceExamples();
}
